File: Python/98. Validate Binary Search Tree.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def isValidBST(self, root: Optional[TreeNode]) -> bool:
#         pre = None
#         def __isValidBST(root):
#             nonlocal pre
#             if not root:
#                 return True
#             left = __isValidBST(root.left)
#             if pre and pre.val >= root.val:
#                 return False
#             pre = root
#             right = __isValidBST(root.right)
#             return left and right
#         return __isValidBST(root)

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        pre = None
        st = [root]
        while st:
            cur = st.pop(-1)
            if cur:
                if cur.right:
                    st.append(cur.right)
                st.append(cur)
                st.append(None)
                if cur.left:
                    st.append(cur.left)
            else:
                cur = st.pop(-1)
                if pre and pre.val >= cur.val:
                    return False
                pre = cur
        return True

File: C++/98. Validate Binary Search Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     TreeNode* pre = NULL;
 //     bool isValidBST(TreeNode* root) {
 //         if (!root) return true;
 //         bool left = isValidBST(root->left);
 //         if (pre && pre->val >= root->val) return false;
 //         pre = root;
 //         bool right = isValidBST(root->right);
 //         return left&&right;
 //     }
 // };

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        stack<TreeNode*> st;
        st.push(root);
        TreeNode* pre = NULL;
        while (!st.empty()) {
            TreeNode* cur = st.top(); st.pop();
            if (cur != NULL) {
                if (cur->right) st.push(cur->right);
                st.push(cur);
                st.push(NULL);
                if (cur->left) st.push(cur->left);
            }
            else {
                TreeNode* cur = st.top(); st.pop();
                if (pre && pre->val >= cur->val) return false;
                pre = cur;
            }
        }
        return true;
    }
};